text/0000-ghost-busting.md

```diff
@@ -0,0 +1,732 @@
+- Feature Name: `ghost_busting`
+- Start Date: 2018-03-04
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+[`PhantomData<T>`]: https://doc.rust-lang.org/nightly/nomicon/phantom-data.html
+[variance]: https://doc.rust-lang.org/nightly/nomicon/subtyping.html
+[drop checking]: https://doc.rust-lang.org/nightly/nomicon/dropck.html
+
+1. Defines unused type parameters to act as if they are logically owned
+   and covariant.
+
+2. Introduces `phantom T` pseudo-fields which improves the ergonomics of
+   changing [variance] and [drop checking] behavior when it is needed.
+
+3. [`PhantomData<T>`] is redefined as `struct PhantomData<T: ?Sized>;`
+   and deprecated.
+
+4. The lang item `phantom_data` is removed.
+
+5. Derive macros for derivable standard library traits will take advantage
+   of statically known phantom types and fields to generate more permissive
+   `impl`s.
+
+# Motivation
+[motivation]: #motivation
+
+## Improving ergonomics and rapid prototyping
+
+Today, it is impossible to define a type with unused type parameters:
+
+```rust
+struct Label<T>;
+```
+
+Instead, you must define the type as:
+
+```rust
+use std::marker::PhantomData;
+
+struct Label<T> {
+    marker: PhantomData<T>,
+}
+```
+
+or equivalently:
+
+```rust
+struct Label<T>(PhantomData<T>);
+```
+
+This is a pain point since users must now:
+
+1. import `std::marker::PhantomData`,
+
+2. add a field of type [`PhantomData<T>`],
+
+3. acknowledge the field by adding an expression `PhantomData`
+   when a value of the type is to be constructed.
+
+These steps add unnecessary boilerplate and is in the way of rapid prototyping
+and developer flow when you are removing and adding fields.
+
+By making `struct Label<T>;` be owning and covariant by default in `T`,
+ergonomics of generics and label types are improved and `PhantomData<T>`
+becomes just a library type.
+
+For types where you really need to change the variance and drop checking
+behavior such as for the following `MyVec<T>` type, we gain in ergonomics
+by being able to skip steps 1 and 3 and instead write:
+
+```rust
+struct MyVec<T> {
+    phantom T,
```

Comment:
Is writing `phantom T` here optional?

Reply:
No, `phantom T` is required to let dropck know that `T` is logically owned, which is not the case if it only sees `data: *const T`. I will clarify.